pkg/runtime/rule/shard.go

```diff
@@ -0,0 +1,132 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package rule
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"hash/crc32"
+	"hash/fnv"
+	"strconv"
+)
+
+import (
+	"github.com/pkg/errors"
+)
+
+import (
+	"github.com/dubbogo/arana/pkg/proto/rule"
+	"github.com/dubbogo/arana/pkg/util/maths"
+)
+
+const (
+	ModShard       ShardType = "modShard"
+	HashMd5Shard   ShardType = "hashMd5Shard"
+	HashCrc32Shard ShardType = "hashCrc32Shard"
+)
+
+var shardMap = map[ShardType]fn{
+	ModShard:       NewModShard,
+	HashMd5Shard:   NewHashMd5Shard,
+	HashCrc32Shard: NewHashCrc32Shard,
+}
+
+func ShardFactory(shardType ShardType, shardNum int) (shardStrategy rule.ShardComputer, err error) {
+	if f, ok := shardMap[shardType]; ok {
+		return f(shardNum), nil
+	}
+	return nil, errors.New("do not have this shardType")
+}
+
+type ShardType string
+type fn func(shardNum int) rule.ShardComputer
+
+type modShard struct {
+	shardNum int
+}
+
+func NewModShard(shardNum int) rule.ShardComputer {
+	return modShard{shardNum}
+}
+
+func (mod modShard) Compute(value interface{}) (int, error) {
+	n, err := strconv.ParseInt(fmt.Sprintf("%v", value), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return int(maths.Abs(n)) % mod.shardNum, nil
+}
+
+// FNV32, FNV32a, FNV64, FNV64a, MD5, SHA1, SHA256, SHA512, murmur3, crc32, crc64, adler
+type hashCrc32Shard struct {
+	shardNum int
+}
+
+func NewHashCrc32Shard(shardNum int) rule.ShardComputer {
+	return hashCrc32Shard{shardNum}
+}
+
+func (h hashCrc32Shard) Compute(value interface{}) (int, error) {
+	s := fmt.Sprintf("%v", value)
+	return int(crc32.ChecksumIEEE([]byte(s))) % h.shardNum, nil
+}
+
+type hashMd5Shard struct {
+	shardNum int
+}
+
+func NewHashMd5Shard(shardNum int) rule.ShardComputer {
+	return hashMd5Shard{shardNum}
+}
+
+func (m hashMd5Shard) Compute(value interface{}) (int, error) {
+	s := fmt.Sprintf("%v", value)
+	md5Hash := md5.New()
+	md5Hash.Write([]byte(s))
+	md5Str := hex.EncodeToString(md5Hash.Sum(nil))
+
+	h := 0
+	byteHash := []byte(md5Str)
+	for i := 0; i < 32; i++ {
+		h <<= 8
+		h |= int(byteHash[i]) & 0xFF
+	}
```

Comment:
The length of md5 bytes is 16, see `md5.Size`, use it directly, I think it's not necessary to convert it as hex string.